--- dgpstep.f	Mon Apr 22 18:17:21 2002
+++ dgpstep_p.f	Mon Apr 22 18:16:59 2002
@@ -57,9 +57,11 @@
 c     rounding errors for the components that are feasible.
 
       do i = 1, n
-         if (x(i) + alpha*w(i) .lt. xl(i)) then
+CORIG         if (x(i) + alpha*w(i) .lt. xl(i)) then
+         if (x(i) + alpha*w(i) .le. xl(i)) then
             s(i) = xl(i) - x(i)
-         else if (x(i) + alpha*w(i) .gt. xu(i)) then
+CORIG         else if (x(i) + alpha*w(i) .gt. xu(i)) then
+         else if (x(i) + alpha*w(i) .ge. xu(i)) then
             s(i) = xu(i) - x(i)
          else
             s(i) = alpha*w(i)
